### PR TITLE
perf/perf_metric.py: add test for CPI STALL json feature

### PR DESCRIPTION
**Pull request by disgoel**
Patch adds a new function to test CPI_STALL_RATIO metricgroup by checking if its present in perf list or not. If metricgroup is present, collect counter statistics using perf stat and check if all stall metrics are present in the output. Also check for multiplexing of events should not happen.

Test for metricgroup events fails when performance information collection property is not enabled in hmc since internally it uses hv_24x7 events to collect statistics. Added a check to cancel the test in such cases.
```
On P10
avocado run --max-parallel-tasks=1 perf_metric.py
JOB ID     : 3266fa2f8fe417f605acf7a5f89d1861de33af5e
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2024-05-24T01.35-3266fa2/job.log
 (1/3) perf_metric.py:perf_metric.test_all_metric_events_with_M: STARTED
 (1/3) perf_metric.py:perf_metric.test_all_metric_events_with_M: PASS (252.43 s)
 (2/3) perf_metric.py:perf_metric.test_all_metric_events_with_metric: STARTED
 (2/3) perf_metric.py:perf_metric.test_all_metric_events_with_metric: PASS (252.33 s)
 (3/3) perf_metric.py:perf_metric.test_cpi_stall: STARTED
 (3/3) perf_metric.py:perf_metric.test_cpi_stall: PASS (0.62 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2024-05-24T01.35-3266fa2/results.html
JOB TIME   : 533.73 s

avocado run --max-parallel-tasks=1 perf_metric.py
JOB ID     : 702a19c3668ce9af0cfe08a5975a80bda839b420
JOB LOG    : /home/disha/avocado-fvt-wrapper/results/job-2024-05-24T12.05-702a19c/job.log
 (1/3) perf_metric.py:perf_metric.test_all_metric_events_with_M: STARTED
 (1/3) perf_metric.py:perf_metric.test_all_metric_events_with_M: CANCEL: Please enable lpar to allow collecting the hv_24x7 counters info (91.93 s)
 (2/3) perf_metric.py:perf_metric.test_all_metric_events_with_metric: STARTED
 (2/3) perf_metric.py:perf_metric.test_all_metric_events_with_metric: CANCEL: Please enable lpar to allow collecting the hv_24x7 counters info (91.92 s)
 (3/3) perf_metric.py:perf_metric.test_cpi_stall: STARTED
 (3/3) perf_metric.py:perf_metric.test_cpi_stall: PASS (1.92 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 2
JOB HTML   : /home/disha/avocado-fvt-wrapper/results/job-2024-05-24T12.05-702a19c/results.html
JOB TIME   : 214.39 s
```
```
On P9
avocado run --max-parallel-tasks=1 perf_metric.py
JOB ID     : 52e85434342fa6c93dc82cf2b6dfc36b019f9583
JOB LOG    : /home/disha/avocado-fvt-wrapper/results/job-2024-05-23T07.57-52e8543/job.log
 (1/1) perf_metric.py:perf_metric.test_cpi_stall: STARTED
 (1/1) perf_metric.py:perf_metric.test_cpi_stall: CANCEL: Feature not supported on platform or metricgroup not present in perf list (0.30 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 1
JOB HTML   : /home/disha/avocado-fvt-wrapper/results/job-2024-05-23T07.57-52e8543/results.html
JOB TIME   : 114.93 s
```
[debug.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/15429552/debug.log)
[debug3.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/15429554/debug3.log)

### CODE REVIEW

**Comment by disgoel** [comment]
@athira-rajeev please review

**Comment by athira-rajeev** [comment]
Hi Disha,

From snippet of result using the CPI stall metric:

```
Performance counter stats for 'system wide':

           1480160      PM_CMPL_STALL                    #     0.25 COMPLETION_STALL_CPI      
            668625      PM_ISSUE_STALL                   #     0.11 ISSUE_STALL_CPI           
           5911750      PM_RUN_INST_CMPL                 #     0.52 DISPATCH_STALL_CPI        
                                                  #     0.93 EXECUTION_STALL_CPI       
           3086333      PM_DISP_STALL_CYC                                                     
           5488256      PM_EXEC_STALL                                                         

       1.241533771 seconds time elapsed
```

the metrics includes: COMPLETION_STALL_CPI, ISSUE_STALL_CPI, DISPATCH_STALL_CPI and EXECUTION_STALL_CPI . Can we check for those also in the test ?

Thanks
Athira

**Comment by disgoel** [comment]
> Hi Disha,
> the metrics includes: COMPLETION_STALL_CPI, ISSUE_STALL_CPI, DISPATCH_STALL_CPI and EXECUTION_STALL_CPI . Can we check for those also in the test ?

@athira-rajeev I have done required changes. Please review.

**Comment by disgoel** [comment]
```
avocado run --max-parallel-tasks=1 perf_metric.py:test_cpi_stall
JOB ID     : 0cf2c969e6c299a2878deeea46e4aa02a9f92fcf
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2024-05-27T20.08-0cf2c96/job.log
 (1/1) perf_metric.py:perf_metric.test_cpi_stall: STARTED
 (1/1) perf_metric.py:perf_metric.test_cpi_stall: PASS (0.41 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2024-05-27T20.08-0cf2c96/results.html
JOB TIME   : 25.40 s
```

File: perf/perf_metric.py [hunk 1]
```diff
@@ -51,6 +51,8 @@ def setUp(self):
         if 'ppc64' not in detected_distro.arch:
             self.cancel("Processor is not PowerPC")
 
+        self.rev = cpu.get_revision()
+
```
Comment (PraveenPenguin) [inline perf/perf_metric.py:55]: Can you please remove this whitespace

File: perf/perf_metric.py [hunk 2]
```diff
@@ -111,6 +116,39 @@ def test_all_metric_events_with_M(self):
...
+            result = (res.stdout + res.stderr).decode()
+            if res.exit_status != 0 or "not counted" in result or "not supported" in result:
+                self.fail("Failed to collect perf stats with CPI_STALL_RATIO"
+                          " metricgroup")
+
+            # Check if the required metrics are present and only once
+            required_metrics = ['PM_CMPL_STALL', 'PM_ISSUE_STALL', 'PM_RUN_INST_CMPL',
+                                'PM_DISP_STALL_CYC', 'PM_EXEC_STALL',
+                                'COMPLETION_STALL_CPI', 'ISSUE_STALL_CPI',
+                                'DISPATCH_STALL_CPI', 'EXECUTION_STALL_CPI']
```
Comment (PraveenPenguin) [inline perf/perf_metric.py:138]: Can these matrices also drive from YAML as if any new Feature added

Comment (disgoel) [reply perf/perf_metric.py:138]: I feel yaml option is not needed here. We will check for these metrics only for this feature.
Still you want I can change code to yaml based.

**Comment by disgoel** [comment]
@PraveenPenguin can you please review

**Review by PraveenPenguin — APPROVED** [review]

**Pull request merged**